### ISSUE TITLE
a bug with Fl::screen_num

### ISSUE THREAD

**Issue by lopo12123** [fltk/fltk#743]
I found this problem while using fltk-rs and reproduced it in fltk (A detailed description of the problem is at https://github.com/fltk-rs/fltk-rs/issues/1422)

---

My code is as follow (The code for reproduction is at https://github.com/MoAlyousef/screen_rep)
``` c++
#include <FL/Fl.H>
#include <stdio.h>
#include <vector>

struct Point {
    int x;
    int y;
    Point(int x_, int y_): x(x_), y(y_) {}
};

int main() {
    for (auto count = 0; count < Fl::screen_count(); count ++) {
        int x = 0, y = 0, w = 0, h = 0;
        Fl::screen_xywh(x, y, w, h);
        printf("%d: (%d, %d, %d, %d)\n", count, x, y, w, h);
    }
    std::vector<Point> points = {
        Point(0, 0),
        Point(0, 1439),
        Point(2559, 0),
        Point(2559, 1439),
        Point(1450, 933),
        Point(1357, 1099),  // x
        Point(1357, 331),
        Point(1323, 1241),  // x
        Point(23, 234),
        Point(23, 1099),
    };
    for (auto point: points) {
        printf("point: %d, %d => screen_num: %d\n", point.x, point.y, Fl::screen_num(point.x, point.y));
    }
}
```

---

The output after running the program is as follows
![image](https://github.com/fltk/fltk/assets/59609929/5c9fc348-d4e1-4c25-8d9c-750f486bd8d7)

---

My system information and cmake version are as follows
![image](https://github.com/fltk/fltk/assets/59609929/2e2a8494-f553-40b6-b18d-f9b3721b28ec)
![image](https://github.com/fltk/fltk/assets/59609929/9793afb5-a6c2-46dc-a097-962350b73130)

**Comment by ManoloFLTK** [comment]
To help you, we need to understand how the 3 displays of your system are positionned relatively to one another. For this, this part of your test program
```
    for (auto count = 0; count < Fl::screen_count(); count ++) {
        int x = 0, y = 0, w = 0, h = 0;
        Fl::screen_xywh(x, y, w, h);
        printf("%d: (%d, %d, %d, %d)\n", count, x, y, w, h);
    }

```
must be  corrected, because as it is presently, the same screen information is repeated 3 times.

Please, replace
`        Fl::screen_xywh(x, y, w, h);`
by
`        Fl::screen_xywh(x, y, w, h, count);`

With this change, the size and location of each of the 3 displays will be output.

Run your program again and post here its output.
TIA

**Comment by lopo12123** [comment]
> To help you, we need to understand how the 3 displays of your system are positionned relatively to one another. For this, this part of your test program
> 
> ```
>     for (auto count = 0; count < Fl::screen_count(); count ++) {
>         int x = 0, y = 0, w = 0, h = 0;
>         Fl::screen_xywh(x, y, w, h);
>         printf("%d: (%d, %d, %d, %d)\n", count, x, y, w, h);
>     }
> ```
> 
> must be corrected, because as it is presently, the same screen information is repeated 3 times.
> 
> Please, replace ` Fl::screen_xywh(x, y, w, h);` by ` Fl::screen_xywh(x, y, w, h, count);`
> 
> With this change, the size and location of each of the 3 displays will be output.
> 
> Run your program again and post here its output. TIA

![image](https://github.com/fltk/fltk/assets/59609929/d3cc9476-9bdf-48ae-b160-db33f2040192)

**Comment by ManoloFLTK** [comment]
Thanks for the corrected data. They indicate you've a display (# 1) of size 2560x1440, another one (# 2) of size 1920x1080 to the right of # 1 and a third, small, display (# 0) of size 1280x720 positioned inside display #1 because the top-left of # 0 is at 424,960 which falls inside # 1.

Is display #0 a copy of display # 1 ? That could explain that one is inside the other.

Anyway, the screen number values returned :
```
0,0 -> 1 OK
0,1439 -> 1 OK
2559,0 -> 1 OK
2559,1439 -> 1 OK
1450,933 -> 1 0K						
1357,1099  -> 0 bad? correct because display 0 is inside 1
1357,331 -> 1 OK
1323,1241  -> 0 bad? correct because display 0 is inside 1
23,234 -> 1  OK
23,1099 -> 1  OK
```
are correct because coordinates 1357,1099 and 1323,1241  are both inside # 0 and # 1.

I would believe your processing of values returned by Fl::screen_num() should take care of the possibility that one coordinate may belong to 2 screens in case of screen copy. It should correct the screen number according to your own logic when Fl::screen_num() is called with coordinates belonging to 2 distinct screens. In that situation, a single coordinate belongs to 2 displays whereas Fl::screen_num() can return only one display number and cannot know which one you expect.

**Comment by lopo12123** [comment]
> Thanks for the corrected data. They indicate you've a display (# 1) of size 2560x1440, another one (# 2) of size 1920x1080 to the right of # 1 and a third, small, display (# 0) of size 1280x720 positioned inside display #1 because the top-left of # 0 is at 424,960 which falls inside # 1.
> 
> Is display #0 a copy of display # 1 ? That could explain that one is inside the other.
> 
> Anyway, the screen number values returned :
> 
> ```
> 0,0 -> 1 OK
> 0,1439 -> 1 OK
> 2559,0 -> 1 OK
> 2559,1439 -> 1 OK
> 1450,933 -> 1 0K						
> 1357,1099  -> 0 bad? correct because display 0 is inside 1
> 1357,331 -> 1 OK
> 1323,1241  -> 0 bad? correct because display 0 is inside 1
> 23,234 -> 1  OK
> 23,1099 -> 1  OK
> ```
> 
> are correct because coordinates 1357,1099 and 1323,1241 are both inside # 0 and # 1.
> 
> I would believe your processing of values returned by Fl::screen_num() should take care of the possibility that one coordinate may belong to 2 screens in case of screen copy. It should correct the screen number according to your own logic when Fl::screen_num() is called with coordinates belonging to 2 distinct screens. In that situation, a single coordinate belongs to 2 displays whereas Fl::screen_num() can return only one display number and cannot know which one you expect.

In fact, my three screens are laid out like this, they are more like an upside-down '品' shape, and the scale_factor of the screen marked 1 on my picture is 1.5. And my three screens are all set to 'extend' instead of 'copy'.
![image](https://github.com/fltk/fltk/assets/59609929/6ee8fd25-1065-4ebd-889b-29d423d4b91a)

I took a photo to prove that screen 1 and screen 2 are indeed extensions and not copy (sorry I had to hide the desktop icons and other details as this is a company device and I don't want to give away some things)
![E955173FC052A98E560F5E36548FAA73](https://github.com/fltk/fltk/assets/59609929/eead49ec-3832-448f-bdcb-c64e676cb215)

Yes, you are right! I tested some boundary points to prove your statement. But then the problem comes to 'Fl::screen_xywh', so it seems that it does not detect my screen 1 information correctly?
![image](https://github.com/fltk/fltk/assets/59609929/9d3d887f-936b-4c80-990e-e7719187583a)

**Comment by ManoloFLTK** [comment]
OK. When the displays in the system contain different scale factor values, the API `Fl::screen_num(int x, int y)` becomes ambiguous and therefore other APIs should be preferred.

The reason is detailed at : https://www.fltk.org/doc-1.4/drawing.html#drawing_DrawingUnit

Thus, for your screen 1 (# 0) Fl::screen_xywh(x, y, w, h, 0) produces values in FLTK units which are drawing units divided by 1.5 :
```
drawing units are 636,1440,1920,1080
FLTK units are 424,960,1280,720
```

API to obtain the display number of an Fl_Window : `int Fl_Window::screen_num()`

API to obtain the scale factor value of a display: `float Fl::screen_scale(int n)`

API to obtain the origin and size of a display : `void Fl::screen_xywh(int &x, int &y, int &w, int &h, int n)`

With these, you should hopefully be able to obtain what you need.

If you don't, please post here what is your exact objective.

**Comment by lopo12123** [comment]
> OK. When the displays in the system contain different scale factor values, the API `Fl::screen_num(int x, int y)` becomes ambiguous and therefore other APIs should be preferred.
> 
> The reason is detailed at : https://www.fltk.org/doc-1.4/drawing.html#drawing_DrawingUnit
> 
> Thus, for your screen 1 (# 0) Fl::screen_xywh(x, y, w, h, 0) produces values in FLTK units which are drawing units divided by 1.5 :
> 
> ```
> drawing units are 636,1440,1920,1080
> FLTK units are 424,960,1280,720
> ```
> 
> API to obtain the display number of an Fl_Window : `int Fl_Window::screen_num()`
> 
> API to obtain the scale factor value of a display: `float Fl::screen_scale(int n)`
> 
> API to obtain the origin and size of a display : `void Fl::screen_xywh(int &x, int &y, int &w, int &h, int n)`
> 
> With these, you should hopefully be able to obtain what you need.
> 
> If you don't, please post here what is your exact objective.

It works fine when used like this. Also when I set all monitors to 'x1' it works fine.

---

I have such a need and don't know how to achieve it.

I'm making a screen capture tool. I have another library that can get images of all screens and their size, position, etc. I want to calculate a large bounding box and use fltk to create a large window covering all screens to support cross-screen screenshots.

The problem now is that when the user uses multiple screens and their zoom ratios are not the same, the bounding box I calculated through fltk sometimes lacks a piece and cannot cover all the screens.

In other words, the requirement is to calculate the size and position of the target screen first when there is no window, and then create a window to cover it, instead of getting its information after already having a window.

**Comment by ManoloFLTK** [comment]
"to calculate the size and position of the target screen first when there is no window" : you already do that with the loop that calls Fl::screen_xywh() over all screens. You must remember that resulting values are in FLTK units, so they are equal to the true screen's pixel values divided by that screen's scale factor.

"create a window to cover it [the target screen]" : the solution is to call  this **before** `Fl_Window::show()` :
`window->screen_num(target_screen);`
where `target_screen` is the number of the target screen (counting from 0). This will force the newly creataed window to the target screen. You may also set your window fullscreen so it covers all the target screen.

```
#include <FL/Fl.H>
#include <FL/Fl_Window.H>

int main() {
  int target_screen = 1; // or 0 or 2
  Fl_Window *full_win = new Fl_Window(100, 100);
  full_win->end();
  full_win->resizable(full_win);
  full_win->screen_num(target_screen);
  full_win->fullscreen();
  full_win->show();
  return Fl::run();
}
```

**Comment by lopo12123** [comment]
> This will force the newly created window to the target screen

'This will force the newly created window to the target screen'

This doesn't seem to be working, or I've misunderstood it. My understanding is that the window will appear on the specified screen, but the reality is that it will only appear on the screen where my terminal is located (no matter how much I specify the target_screen)

**Comment by ManoloFLTK** [comment]
Sorry. I have forgotten an ingredient in the recipe. It's also necessary to set the expected position of the window before calling show() for the window. That can be done either using the 5-argument Fl_Window constructor, or calling Fl_Window->position(int X, int Y) before calling show(). Both ways are equally correct.

With this, you should be good.
```
#include <FL/Fl.H>
#include <FL/Fl_Window.H>

int main() {
  int target_screen = 0; // or 1 or 2
  int X,Y,W,H;
  Fl::screen_xywh(X,Y,W,H,target_screen);
  Fl_Window *full_win = new Fl_Window(/*X, Y,*/ 200, 200);
  full_win->end();
  full_win->resizable(full_win);
  full_win->screen_num(target_screen);
  full_win->position(X, Y);
  full_win->fullscreen();
  full_win->show();
  return Fl::run();
}
```

**Comment by lopo12123** [comment]
> 



> Sorry. I have forgotten an ingredient in the recipe. It's also necessary to set the expected position of the window before calling show() for the window. That can be done either using the 5-argument Fl_Window constructor, or calling Fl_Window->position(int X, int Y) before calling show(). Both ways are equally correct.
> 
> With this, you should be good.
> 
> ```
> #include <FL/Fl.H>
> #include <FL/Fl_Window.H>
> 
> int main() {
>   int target_screen = 0; // or 1 or 2
>   int X,Y,W,H;
>   Fl::screen_xywh(X,Y,W,H,target_screen);
>   Fl_Window *full_win = new Fl_Window(/*X, Y,*/ 200, 200);
>   full_win->end();
>   full_win->resizable(full_win);
>   full_win->screen_num(target_screen);
>   full_win->position(X, Y);
>   full_win->fullscreen();
>   full_win->show();
>   return Fl::run();
> }
> ```

Thanks for your patience, now the window is properly fullscreen on the target screen!

---

I tried to comment out this line: `full_win->fullscreen()`, and found that the position where the window appeared was a bit weird. Is this because the zoom ratio of one of my screens is x1.5?

![image](https://github.com/fltk/fltk/assets/59609929/35ec3930-ba64-4200-8e0d-4bd263376269)
![image](https://github.com/fltk/fltk/assets/59609929/7c74f19b-f19e-41ac-862d-39be3a2a909c)
![image](https://github.com/fltk/fltk/assets/59609929/bc5b4598-544f-4b96-88e2-ec223904345a)

And when the window is expected to appear on screen 0, because of a positioning problem, it flickers a bit and then changes to the correct size on screen 1.

I don’t have screen recording software on my computer, so the shots are blurry, but you should still be able to see the flickering phenomenon

https://github.com/fltk/fltk/assets/59609929/4fe6b9ce-d221-46cf-8b7b-f17b21c5751e

**Comment by ManoloFLTK** [comment]
OK. It seems that the call to Fl_Window::position() is not working well. Please, use the approach with the 5-argument Fl_Window constructor. In my test here with 2 screens and one scaled at 1.5 it's OK.
```
#include <FL/Fl.H>
#include <FL/Fl_Window.H>

int main(int argc, char **argv) {
  int target_screen = 1; // or 0 or 2
  int X,Y,W,H;
  Fl::screen_xywh(X,Y,W,H,target_screen);
  fprintf(stderr,"screen no=%d s=%.1f %d,%d\n",
          target_screen, Fl::screen_scale(target_screen),X,Y);
  Fl_Window *full_win = new Fl_Window(X, Y, 200, 200);
  full_win->end();
  full_win->resizable(full_win);
  full_win->screen_num(target_screen);
  //full_win->position(X, Y);
  //full_win->fullscreen();
  full_win->show();
  return Fl::run();
}
```

**Comment by lopo12123** [comment]
> I was wrong before when I stated that 2 approaches are possible to define the window position in a screen when there are several screens. Only the 5-argument Fl_Window constructor is efficient. The correct procedure is given below :
> 
> ```
> #include <FL/Fl.H>
> #include <FL/Fl_Window.H>
> 
> int main(int argc, char **argv) {
>   int target_screen = 1; // or 0 or 2
>   int X,Y,W,H;
>   Fl::screen_xywh(X,Y,W,H,target_screen);
>   Fl_Window *full_win = new Fl_Window(X, Y, 200, 200);
>   full_win->end();
>   full_win->resizable(full_win);
>   full_win->screen_num(target_screen);
>   //full_win->fullscreen();
>   full_win->show();
>   return Fl::run();
> }
> ```

1. When I use your code to set target_screen to 0 it still has the previous error. (correct when target_screen is 1 or 2)

2. However I temporarily moved the screen 1 to the left so the three screens are horizontal and your code works fine. (correct when target_screen is 0 or 1 or 2)
![9B2C8Z5W_FXV)5J@17IUYBI](https://github.com/fltk/fltk/assets/59609929/f8101cbb-60e0-4e48-bfd0-0a6fa2f6584b)

3. Then I set the zoom ratio of screen 3 to x1.25 for testing, the window is normal on screen 0 and 1, but misaligned on screen 2. (Correct when target_screen is 0 or 1)
![YWJ_BGCFYS2@14I}}$C6U4P](https://github.com/fltk/fltk/assets/59609929/53cb6fe3-59ac-4a9f-9780-f68782107ab3)

So I wonder if the relative position between the screens also has a certain influence.

--- 

I'm guessing maybe a multi-screen test is needed to find out how it works (test with different combinations of screen layouts and zooms). I don't know c++ language, maybe the logic of fltk itself is enough to explain

**Comment by ManoloFLTK** [comment]
Did you perform the experiments you report above with the last git commit of FLTK? Or did you do it via fltk-rs? I'm asking to be sure you used the last (or very recent) FLTK 1.4 source code.

**Comment by lopo12123** [comment]
> Did you perform the experiments you report above with the last git commit of FLTK? Or did you do it via fltk-rs? I'm asking to be sure you used the last (or very recent) FLTK 1.4 source code.

My cmakelists file is as follows, I don't know much about it, but it seems to use the code on the master branch. If I have any mistakes, please help me to modify it and I will test again.

![image](https://github.com/fltk/fltk/assets/59609929/90325fc5-deb9-4c08-ba55-ab399213e34c)

**Comment by ManoloFLTK** [comment]
OK. Your CMake file shows you do use the last FLTK source code. That's good.

I tried here with 3 screens arranged with 2 above and one scaled at 1.25 below and got no error.

Can you run the test app from your terminal? If yes, please use the code below
```
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Button.H>

int main(int argc, char **argv) {
  int X,Y,W,H;
  Fl::screen_xywh(X,Y,W,H,0);
  printf("0: %dx%d - %dx%d %.1f\n",X,Y,W,H,Fl::screen_scale(0));
  Fl::screen_xywh(X,Y,W,H,1);
  printf("1: %dx%d - %dx%d %.1f\n",X,Y,W,H,Fl::screen_scale(1));
  Fl::screen_xywh(X,Y,W,H,2);
  printf("2: %dx%d - %dx%d %.1f\n",X,Y,W,H,Fl::screen_scale(2));

  int target_screen = 0; //  0 or 1 or 2
  if (argc == 2) {
    sscanf(argv[1], "%d", &target_screen);
  }
  Fl::screen_xywh(X,Y,W,H,target_screen);
  Fl_Window *full_win = new Fl_Window(X, Y, 200, 200);
  Fl_Button b(50,50,50,30, "OK");
  full_win->end();
  full_win->resizable(full_win);
  full_win->screen_num(target_screen);
  //full_win->fullscreen();
  full_win->show();
  return Fl::run();
}
```
and report the output it gives in your terminal. In my test here with 3 displays, I obtain:
```
0: 0x0 - 1075x669 1.3
1: -276x-768 - 1024x768 1.0
2: 748x-768 - 1024x768 1.0
```

With this program, it's no longer necessary to rebuild it to change the target screen: just give the desired screen number as an argument of the command line:
```
./test 0
or
./test 2
```

Please, try this and report the outputs in 2 screen arrangements, one where there's a problem and one where there's none.

**Comment by lopo12123** [comment]
> OK. Your CMake file shows you do use the last FLTK source code. That's good.
> 
> I tried here with 3 screens arranged with 2 above and one scaled at 1.25 below and got no error.
> 
> Can you run the test app from your terminal? If yes, please use the code below
> 
> ```
> #include <FL/Fl.H>
> #include <FL/Fl_Window.H>
> #include <FL/Fl_Button.H>
> 
> int main(int argc, char **argv) {
>   int X,Y,W,H;
>   Fl::screen_xywh(X,Y,W,H,0);
>   printf("0: %dx%d - %dx%d %.1f\n",X,Y,W,H,Fl::screen_scale(0));
>   Fl::screen_xywh(X,Y,W,H,1);
>   printf("1: %dx%d - %dx%d %.1f\n",X,Y,W,H,Fl::screen_scale(1));
>   Fl::screen_xywh(X,Y,W,H,2);
>   printf("2: %dx%d - %dx%d %.1f\n",X,Y,W,H,Fl::screen_scale(2));
> 
>   int target_screen = 0; //  0 or 1 or 2
>   if (argc == 2) {
>     sscanf(argv[1], "%d", &target_screen);
>   }
>   Fl::screen_xywh(X,Y,W,H,target_screen);
>   Fl_Window *full_win = new Fl_Window(X, Y, 200, 200);
>   Fl_Button b(50,50,50,30, "OK");
>   full_win->end();
>   full_win->resizable(full_win);
>   full_win->screen_num(target_screen);
>   //full_win->fullscreen();
>   full_win->show();
>   return Fl::run();
> }
> ```
> 
> and report the output it gives in your terminal. In my test here with 3 displays, I obtain:
> 
> ```
> 0: 0x0 - 1075x669 1.3
> 1: -276x-768 - 1024x768 1.0
> 2: 748x-768 - 1024x768 1.0
> ```
> 
> With this program, it's no longer necessary to rebuild it to change the target screen: just give the desired screen number as an argument of the command line:
> 
> ```
> ./test 0
> or
> ./test 2
> ```
> 
> Please, try this and report the outputs in 2 screen arrangements, one where there's a problem and one where there's none.

I ran your code and the output for different layout cases is as follows

1. 
![image](https://github.com/fltk/fltk/assets/59609929/392aa605-6afd-4d9b-a7e1-36c10419d8d2)
![image](https://github.com/fltk/fltk/assets/59609929/b6baf495-8e6d-44f8-a831-cb66ff2ea290)

2.
![image](https://github.com/fltk/fltk/assets/59609929/8c7849a7-9c00-4342-844e-bdae4f873d83)
![image](https://github.com/fltk/fltk/assets/59609929/770b92f1-6b1f-43cb-97d5-798be6fe3f70)

3.
![image](https://github.com/fltk/fltk/assets/59609929/f1b51537-0ce9-4275-af66-529c3dc36400)
![image](https://github.com/fltk/fltk/assets/59609929/614c56de-699a-42d1-a8d0-cebaaa911f47)

4.
![image](https://github.com/fltk/fltk/assets/59609929/581a1a25-9297-4f50-8e03-5613f85f90ac)
![image](https://github.com/fltk/fltk/assets/59609929/665133f7-de0f-4887-b0d4-91c5178b67ad)

**Comment by ManoloFLTK** [comment]
Thanks for these data. I see nothing that could explain your problems.

Could you, please, try with this program?
It should open a window centered on each of your screens (you'll have to close them).
Just type its name to run it. No screen number needed in the command line.

If you uncomment the ` //full_win->fullscreen();` statement, you should obtain one fullscreen window on each of your screens.

```
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Button.H>

int main() {
  int X,Y,W,H, nscreens = Fl::screen_count();
  for (int i = 0; i < nscreens; i++) {
    Fl::screen_xywh(X,Y,W,H,i);
    printf("%d: %dx%d - %dx%d %.1f\n",i,X,Y,W,H,Fl::screen_scale(i));
  }

  for (int target_screen = 0; target_screen < nscreens; target_screen++) {
    Fl::screen_xywh(X,Y,W,H,target_screen);
    Fl_Window *full_win = new Fl_Window(X + W/2 - 100, Y + H/2 - 100, 200, 200);
    new Fl_Button(50,50,50,30, "OK");
    full_win->end();
    full_win->resizable(full_win);
    //full_win->border(0);
    full_win->screen_num(target_screen);
    //full_win->fullscreen();
    full_win->show();
  }
  return Fl::run();
}
```

**Comment by ManoloFLTK** [comment]
Very well. Then it might just be because targetting the exact top-left angle of a display was a bit risky because of possible rounding errors after multiplying with scale factors (1.5, 1.25).

Let's wrap up.

To control where to put a window in multiple screens
1) get with `Fl::screen_xywh(X,Y,W,H, n) `the origin and size, in FLTK units, of screen # n (counting screens from 0)
2) construct your window with the 5-argument Fl_Window constructor
`Fl_Window(x, y, w, h, "title")`
where x,y is the top-left and w,h the window size
3) force the screen number with
`window->screen_num(n)`
4) optionally, if you want your window fullscreen, add
`window->fullscreen()`

**Comment by ManoloFLTK** [comment]
> I tried the previous code again and it does seem to misposition a lot less when I set the window's position not so close to the edge. 

The x,y coordinates given to the `Fl_Window` constructor represent the screen position of the client part of the window (that is, the titlebar and window edges are excluded). You therefore have to account for that to make sure the window and its titlebar fall inside the target screen area.

@lopo12123 Please, close issue if you have no more questions.
